public/app/features/dashboard/state/DashboardModel.ts

```diff
@@ -756,59 +760,56 @@ export class DashboardModel implements TimeModel {
 
-      setScopedVars(rowCopy, option);
+      setScopedVars(rowClone, variable, curOption);
 
-      const rowHeight = this.getRowHeight(rowCopy);
-      const rowPanels = rowCopy.panels || [];
+      const rowHeight = this.getRowHeight(rowClone);
+      const panelsInRow = rowClone.panels || [];
       let panelBelowIndex;
 
       if (panel.collapsed) {
-        // For collapsed row just copy its panels and set scoped vars and proper IDs
-        for (const rowPanel of rowPanels) {
-          setScopedVars(rowPanel, option);
+        // For a collapsed row, just copy its panels, set scoped vars and proper IDs
+        for (const panelInRow of panelsInRow) {
+          setScopedVars(panelInRow, variable, curOption);
           if (optionIndex > 0) {
-            this.updateRepeatedPanelIds(rowPanel, true);
+            this.updateRepeatedPanelIds(panelInRow, true);
           }
         }
-        rowCopy.gridPos.y += optionIndex;
-        yPos += optionIndex;
+
+        // push nth row clone's y-pos down by n
+        rowClone.gridPos.y += optionIndex;
         panelBelowIndex = panelIndex + optionIndex + 1;
       } else {
-        // insert after 'row' panel
-        const insertPos = panelIndex + (rowPanels.length + 1) * optionIndex + 1;
-        rowPanels.forEach((rowPanel: PanelModel, i: number) => {
-          setScopedVars(rowPanel, option);
+        // insert after row panel
+        const insertPos = panelIndex + (panelsInRow.length + 1) * optionIndex + 1;
+        panelsInRow.forEach((panelInRow, i) => {
+          setScopedVars(panelInRow, variable, curOption);
+
           if (optionIndex > 0) {
-            const cloneRowPanel = new PanelModel(rowPanel);
-            this.updateRepeatedPanelIds(cloneRowPanel, true);
-            // For exposed row additionally set proper Y grid position and add it to dashboard panels
-            cloneRowPanel.gridPos.y += rowHeight * optionIndex;
-            this.panels.splice(insertPos + i, 0, cloneRowPanel);
+            const panelInRowClone = new PanelModel(panelInRow);
+            this.updateRepeatedPanelIds(panelInRowClone, true);
+
+            // For exposed row, set correct grid y-position and add it to dashboard panels
+            panelInRowClone.gridPos.y += rowHeight * optionIndex;
+            this.panels.splice(insertPos + i, 0, panelInRowClone);
           }
         });
-        rowCopy.panels = [];
-        rowCopy.gridPos.y += rowHeight * optionIndex;
-        yPos += rowHeight;
-        panelBelowIndex = insertPos + rowPanels.length;
+
+        rowClone.panels = [];
+        rowClone.gridPos.y += rowHeight * optionIndex;
+        panelBelowIndex = insertPos + panelsInRow.length;
       }
 
       // Update gridPos for panels below if we inserted more than 1 repeated row panel
       if (selectedOptions.length > 1) {
         for (const panel of this.panels.slice(panelBelowIndex)) {
-          panel.gridPos.y += yPos;
+          panel.gridPos.y += rowHeight;
```

Comment:
This was the main cause of the issue. Basically the y-pos of panels following a row was increased much more than needed, and for dashboard's with many rows, this could cause `panel.gridPos.y` to reach the max safe integer value, causing the panels to all sort of bunch up at the end.